File: src/main/java/dev/nytweetdeck/post/TranslationMemory.java
package dev.nytweetdeck.post;

import java.util.LinkedHashMap;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.function.Predicate;
import java.util.function.Supplier;

/** Bounded successful results and unbounded-by-LRU pending requests, scoped to this process. */
public final class TranslationMemory<T> {
    private final int capacity;
    private final Map<Key, T> values = new LinkedHashMap<>(128, 0.75f, true);
    private final Map<Key, CompletableFuture<T>> pending = new HashMap<>();

    public TranslationMemory(int capacity) {
        this.capacity = capacity;
    }

    public T load(Key key, Supplier<T> operation, Predicate<T> successful) {
        return load(key, operation, successful, () -> {}, () -> {});
    }

    public T load(Key key, Supplier<T> operation, Predicate<T> successful,
            Runnable onCacheHit, Runnable onJoined) {
        CompletableFuture<T> request;
        boolean owner;
        synchronized (this) {
            var cached = values.get(key);
            if (cached != null) {
                onCacheHit.run();
                return cached;
            }
            request = pending.get(key);
            owner = request == null;
            if (owner) {
                request = new CompletableFuture<>();
                pending.put(key, request);
            }
        }
        if (!owner) {
            onJoined.run();
            try {
                return request.join();
            } catch (CompletionException exception) {
                if (exception.getCause() instanceof RuntimeException runtime) throw runtime;
                throw exception;
            }
        }
        try {
            var value = operation.get();
            synchronized (this) {
                if (successful.test(value)) {
                    values.put(key, value);
                    if (values.size() > capacity) values.remove(values.keySet().iterator().next());
                }
            }
            request.complete(value);
            return value;
        } catch (RuntimeException | Error failure) {
            request.completeExceptionally(failure);
            throw failure;
        } finally {
            synchronized (this) { pending.remove(key, request); }
        }
    }

    public synchronized int size() { return values.size(); }
    public synchronized int pendingCount() { return pending.size(); }

    public record Key(String accountId, String kind, String id, String sourceLanguage,
            String targetLanguage, String text) {}
}

File: android/app/src/main/java/dev/nytweetdeck/android/data/TranslationMemory.kt
package dev.nytweetdeck.android.data

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException

/** Process-local, bounded successful results. Concurrent readers share only the same request. */
internal class TranslationMemory<K, V>(private val capacity: Int) {
    private val values = LinkedHashMap<K, V>(16, 0.75f, true)
    private val pending = mutableMapOf<K, CompletableFuture<V>>()

    init { require(capacity > 0) }

    val size: Int get() = synchronized(this) { values.size }
    val inFlight: Int get() = synchronized(this) { pending.size }

    fun getOrLoad(
        key: K,
        cacheable: (V) -> Boolean = { true },
        onHit: () -> Unit = {},
        onJoin: () -> Unit = {},
        load: () -> V,
    ): V {
        val future: CompletableFuture<V>
        val owner: Boolean
        synchronized(this) {
            if (values.containsKey(key)) {
                onHit()
                return values.getValue(key)
            }
            val existing = pending[key]
            owner = existing == null
            future = existing ?: CompletableFuture<V>().also { pending[key] = it }
        }
        if (!owner) {
            onJoin()
            return try { future.join() } catch (exception: CompletionException) {
                throw exception.cause ?: exception
            }
        }
        try {
            val value = load()
            synchronized(this) {
                if (cacheable(value)) {
                    values[key] = value
                    while (values.size > capacity) {
                        val iterator = values.entries.iterator()
                        iterator.next()
                        iterator.remove()
                    }
                }
                future.complete(value)
                pending.remove(key)
            }
            return value
        } catch (failure: Throwable) {
            synchronized(this) {
                future.completeExceptionally(failure)
                pending.remove(key)
            }
            throw failure
        }
    }
}
